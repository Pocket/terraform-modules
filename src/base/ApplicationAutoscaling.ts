import { Resource } from 'cdktf';
import { appautoscaling, cloudwatch, iam } from '@cdktf/provider-aws';
const { AppautoscalingPolicy, AppautoscalingTarget } = appautoscaling;
const { CloudwatchMetricAlarm } = cloudwatch;
const { IamRole, IamRolePolicy, DataAwsIamPolicyDocument } = iam;
import { Construct } from 'constructs';

export interface ApplicationAutoscalingProps {
  ecsClusterName: string;
  ecsServiceName: string;
  prefix: string;
  scalableDimension: string;
  scaleInThreshold: number;
  scaleOutThreshold: number;
  stepScaleInAdjustment: number;
  stepScaleOutAdjustment: number;
  tags?: { [key: string]: string };
  targetMaxCapacity: number;
  targetMinCapacity: number;
}

/*
 * Generates an AutoScaling group
 */

export class ApplicationAutoscaling extends Resource {
  constructor(
    scope: Construct,
    name: string,
    config: ApplicationAutoscalingProps
  ) {
    super(scope, name);

    // set up IAM role & policy
    const iamRole = ApplicationAutoscaling.generateIamRole(this, config);
    ApplicationAutoscaling.generateIamRolePolicy(this, config, iamRole);

    // set up autoscaling target & in/out policies
    const autoScalingTarget = ApplicationAutoscaling.generateAutoScalingTarget(
      this,
      config,
      iamRole
    );

    const applicationScaleOut =
      ApplicationAutoscaling.generateAutoSclaingPolicy(
        this,
        config,
        autoScalingTarget,
        'Out'
      );

    const applicationScaleIn = ApplicationAutoscaling.generateAutoSclaingPolicy(
      this,
      config,
      autoScalingTarget,
      'In'
    );

    // set up cloudwatch alarms
    ApplicationAutoscaling.generateCloudwatchMetricAlarm(
      this,
      config,
      'scale_out_alarm',
      `${config.prefix} Service High CPU`,
      'Alarm to add capacity if container CPU is high',
      'GreaterThanThreshold',
      config.scaleOutThreshold,
      applicationScaleOut.arn
    );

    ApplicationAutoscaling.generateCloudwatchMetricAlarm(
      this,
      config,
      'scale_in_alarm',
      `${config.prefix} Service Low CPU`,
      'Alarm to reduce capacity if container CPU is low',
      'LessThanThreshold',
      config.scaleInThreshold,
      applicationScaleIn.arn
    );
  }

  /**
   * Creates an IAM Role
   * @param resource
   * @param config
   * @returns IamRole
   */
  static generateIamRole(
    resource: Resource,
    config: ApplicationAutoscalingProps
  ): iam.IamRole {
    return new IamRole(resource, `autoscaling_role`, {
      name: `${config.prefix}-AutoScalingRole`,
      assumeRolePolicy: new DataAwsIamPolicyDocument(
        resource,
        `autoscaling_assume`,
        {
          statement: [
            {
              effect: 'Allow',
              actions: ['sts:AssumeRole'],
              principals: [
                {
                  identifiers: ['ecs.application-autoscaling.amazonaws.com'],
                  type: 'Service',
                },
              ],
            },
          ],
        }
      ).json,
    });
  }

  /**
   * Creates an IAM Role Policy
   *
   * @param resource
   * @param config
   * @param iamRole
   */
  static generateIamRolePolicy(
    resource: Resource,
    config: ApplicationAutoscalingProps,
    iamRole: iam.IamRole
  ): void {
    new IamRolePolicy(resource, `autoscaling_role_policy`, {
      name: `${config.prefix}-AutoScalingPolicy`,
      role: iamRole.id,
      policy: new DataAwsIamPolicyDocument(resource, `role_policy`, {
        statement: [
          {
            effect: 'Allow',
            actions: [
              'cloudwatch:PutMetricAlarm',
              'cloudwatch:DescribeAlarms',
              'cloudwatch:DeleteAlarms',
            ],
            resources: [
              `arn:aws:cloudwatch:*:*:alarm:/${config.ecsServiceName}*`,
            ],
          },
          {
            effect: 'Allow',
            actions: ['ecs:DescribeServices', 'ecs:UpdateService'],
            resources: [`arn:aws:ecs:*:*:service/${config.ecsServiceName}*`],
          },
        ],
      }).json,
    });
  }

  /**
   * Creates an Auto Scaling Target
   * @param resource
   * @param config
   * @param iamRole
   * @returns AppautoscalingTarget
   */
  static generateAutoScalingTarget(
    resource: Resource,
    config: ApplicationAutoscalingProps,
    iamRole: iam.IamRole
  ): appautoscaling.AppautoscalingTarget {
    return new AppautoscalingTarget(resource, `autoscaling_target`, {
      maxCapacity: config.targetMaxCapacity,
      minCapacity: config.targetMinCapacity,
      resourceId: `service/${config.ecsClusterName}/${config.ecsServiceName}`,
      roleArn: iamRole.arn,
      scalableDimension: 'ecs:service:DesiredCount',
      serviceNamespace: 'ecs',
    });
  }

  /**
   * Creates an Autoscaling Policy
   * @param resource
   * @param config
   * @param target
   * @param type
   * @returns AppautoscalingPolicy
   */
  static generateAutoSclaingPolicy(
    resource: Resource,
    config: ApplicationAutoscalingProps,
    target: appautoscaling.AppautoscalingTarget,
    type: 'In' | 'Out'
  ): appautoscaling.AppautoscalingPolicy {
    let stepAdjustment;

    if (type === 'In') {
      stepAdjustment = [
        {
          metricIntervalUpperBound: '0',
          scalingAdjustment: config.stepScaleInAdjustment,
        },
      ];
    } else {
      stepAdjustment = [
        {
          metricIntervalLowerBound: '0',
          scalingAdjustment: config.stepScaleOutAdjustment,
        },
      ];
    }

    return new AppautoscalingPolicy(
      resource,
      `scale_${type.toLowerCase()}_policy`,
      {
        name: `${config.prefix}-Scale${type}Policy`,
        policyType: 'StepScaling',
        resourceId: target.resourceId,
        scalableDimension: target.scalableDimension,
        serviceNamespace: target.serviceNamespace,

        stepScalingPolicyConfiguration: {
          adjustmentType: `ChangeInCapacity`,
          cooldown: 60,
          metricAggregationType: 'Average',
          stepAdjustment,
        },
        dependsOn: [target],
      }
    );
  }

  /**
   * Creates a Cloudwatch Metric Alarm
   * @param resource
   * @param config
   * @param id
   * @param name
   * @param desc
   * @param operator
   * @param threshold
   * @param arn
   */
  static generateCloudwatchMetricAlarm(
    resource: Resource,
    config: ApplicationAutoscalingProps,
    id: string,
    name: string,
    desc: string,
    operator: string,
    threshold: number,
    arn: string
  ): void {
    new CloudwatchMetricAlarm(resource, id, {
      alarmName: name,
      alarmDescription: desc,
      comparisonOperator: operator,
      evaluationPeriods: 2,
      threshold,
      statistic: 'Average',
      period: 60,
      namespace: 'AWS/ECS',
      metricName: 'CPUUtilization',
      treatMissingData: 'notBreaching',
      dimensions: {
        ClusterName: config.ecsClusterName,
        ServiceName: config.ecsServiceName,
      },
      alarmActions: [arn],
      tags: config.tags,
    });
  }
}
