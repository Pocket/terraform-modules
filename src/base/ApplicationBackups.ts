import { Construct } from 'constructs';
import { TerraformMetaArguments } from 'cdktf';
import {
  BackupPlanRule,
  BackupPlan,
} from '@cdktf/provider-aws/lib/backup-plan';
import {
  BackupSelection,
  BackupSelectionSelectionTag,
} from '@cdktf/provider-aws/lib/backup-selection';
import { BackupVault } from '@cdktf/provider-aws/lib/backup-vault';
import { BackupVaultPolicy } from '@cdktf/provider-aws/lib/backup-vault-policy';

export interface ApplicationBackupProps extends TerraformMetaArguments {
  name: string;
  kmsKeyArn: string;
  prefix: string;
  accountId: string;
  vaultPolicy: string;
  backupPlans: {
    name: string;
    resources: string[];
    rules: Omit<BackupPlanRule, 'targetVaultName'>[];
    selectionTag: BackupSelectionSelectionTag[];
  }[];
  tags?: { [key: string]: string };
}

export class ApplicationBackup extends Construct {
  public backupPlan: BackupPlan;
  public backupSelection: BackupSelection;
  public backupPlanRule: BackupPlanRule;
  private static vault: BackupVault;

  constructor(
    scope: Construct,
    name: string,
    private config: ApplicationBackupProps,
  ) {
    super(scope, name);

    const vault = new BackupVault(this, 'backup-vault', {
      name: `${config.prefix}-${config.name}`,
      kmsKeyArn: config.kmsKeyArn,
      tags: config.tags,
      provider: config.provider,
    });

    new BackupVaultPolicy(this, 'backup-vault-policy', {
      backupVaultName: vault.name,
      policy: config.vaultPolicy,
      provider: config.provider,
    });

    config.backupPlans.forEach((plan) => {
      const backupPlan = new BackupPlan(this, 'backup-plan', {
        name: plan.name,
        rule: plan.rules.map((rule) => ({
          ...rule,
          targetVaultName: vault.name,
        })),
        tags: config.tags,
        provider: config.provider,
      });

      new BackupSelection(this, 'backup-selection', {
        name: `${config.prefix}-Backup-Selection`,
        planId: backupPlan.id,
        iamRoleArn: `arn:aws:iam::${config.accountId}:role/service-role/AWSBackupDefaultServiceRole`,
        resources: plan.resources,
        selectionTag: plan.selectionTag,
        provider: config.provider,
      });
    });
  }
}
