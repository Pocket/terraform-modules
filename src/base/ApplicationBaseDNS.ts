import { Resource } from 'cdktf';
import {
  DataAwsRoute53Zone,
  Route53Record,
  Route53Zone,
} from '../../.gen/providers/aws';
import { Construct } from 'constructs';
import { getRootDomain } from '../utilities';

export interface RootDNSProps {
  domain: string;
  tags?: { [key: string]: string };
}

export class ApplicationBaseDNS extends Resource {
  public readonly zoneId: string;

  constructor(scope: Construct, name: string, config: RootDNSProps) {
    super(scope, name);

    const route53MainZone = new DataAwsRoute53Zone(
      scope,
      `${name}_main_hosted_zone`,
      {
        name: getRootDomain(config.domain),
        tags: config.tags,
      }
    );

    const route53SubZone = new Route53Zone(this, `${name}_subhosted_zone`, {
      name: config.domain,
      tags: config.tags,
    });

    this.zoneId = route53SubZone.zoneId;

    new Route53Record(this, `${name}_subhosted_zone_ns`, {
      name: config.domain,
      type: 'NS',
      ttl: 86400,
      zoneId: route53MainZone.zoneId,
      records: route53SubZone.nameServers,
    });
  }
}
