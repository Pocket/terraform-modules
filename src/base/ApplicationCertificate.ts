import { AcmCertificate } from '@cdktf/provider-aws/lib/acm-certificate';
import { AcmCertificateValidation } from '@cdktf/provider-aws/lib/acm-certificate-validation';
import { DataAwsRoute53Zone } from '@cdktf/provider-aws/lib/data-aws-route53-zone';
import { Route53Record } from '@cdktf/provider-aws/lib/route53-record';
import { TerraformMetaArguments, TerraformProvider } from 'cdktf';
import { Construct } from 'constructs';

export interface ApplicationCertificateProps extends TerraformMetaArguments {
  domain: string;
  /**
   * If zoneId is not passed then we use a data block and the zoneDomain to grab it.
   */
  zoneId?: string;
  zoneDomain?: string;
  tags?: { [key: string]: string };
}

/**
 * Generates an Application Certificate given a domain name and zoneId
 */
export class ApplicationCertificate extends Construct {
  public readonly arn: string;
  // Use `certificateValidation` in `dependsOn` to block on the
  // complete certificate for any downstream dependencies
  public readonly certificateValidation: AcmCertificateValidation;

  constructor(
    scope: Construct,
    name: string,
    config: ApplicationCertificateProps
  ) {
    super(scope, name);

    if (!config.zoneId && config.zoneDomain) {
      const route53Zone = new DataAwsRoute53Zone(this, `zone`, {
        name: config.zoneDomain,
        provider: config.provider,
      });
      config.zoneId = route53Zone.zoneId;
    } else if (!config.zoneId && !config.zoneDomain) {
      throw new Error('You need to pass either a zone id or a zone domain');
    }

    const certificate = ApplicationCertificate.generateAcmCertificate(
      this,
      config.domain,
      config.tags,
      config.provider
    );

    const certificateRecord = ApplicationCertificate.generateRoute53Record(
      this,
      config.zoneId,
      certificate,
      config.provider
    );

    const validation = ApplicationCertificate.generateAcmCertificateValidation(
      this,
      certificate,
      certificateRecord,
      config.provider
    );

    this.arn = certificate.arn;
    this.certificateValidation = validation;
  }

  static generateAcmCertificate(
    scope: Construct,
    domain: string,
    tags?: { [key: string]: string },
    provider?: TerraformProvider
  ): AcmCertificate {
    return new AcmCertificate(scope, `certificate`, {
      domainName: domain,
      validationMethod: 'DNS',
      tags: tags,
      lifecycle: {
        createBeforeDestroy: true,
      },
      provider: provider,
    });
  }

  static generateRoute53Record(
    scope: Construct,
    zoneId: string,
    cert: AcmCertificate,
    provider?: TerraformProvider
  ): Route53Record {
    const record = new Route53Record(scope, `certificate_record`, {
      name: cert.domainValidationOptions.get(0).resourceRecordName,
      type: cert.domainValidationOptions.get(0).resourceRecordType,
      zoneId,
      records: [cert.domainValidationOptions.get(0).resourceRecordValue],
      ttl: 60,
      dependsOn: [cert],
      provider: provider,
    });

    return record;
  }

  static generateAcmCertificateValidation(
    scope: Construct,
    cert: AcmCertificate,
    record: Route53Record,
    provider?: TerraformProvider
  ): AcmCertificateValidation {
    return new AcmCertificateValidation(scope, `certificate_validation`, {
      certificateArn: cert.arn,
      validationRecordFqdns: [record.fqdn],
      dependsOn: [record, cert],
      provider: provider,
    });
  }
}
