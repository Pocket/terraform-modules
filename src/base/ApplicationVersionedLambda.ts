import { Fn, Resource } from 'cdktf';
import { Construct } from 'constructs';
import { cloudwatch, iam, lambdafunction, s3 } from '@cdktf/provider-aws';
import {
  DataArchiveFile,
  DataArchiveFileSource,
} from '@cdktf/provider-archive';

export enum LAMBDA_RUNTIMES {
  PYTHON38 = 'python3.8',
  NODEJS12 = 'nodejs12.x',
  NODEJS14 = 'nodejs14.x',
}

export interface ApplicationVersionedLambdaProps {
  name: string;
  description?: string;
  runtime: LAMBDA_RUNTIMES;
  handler: string;
  timeout?: number;
  environment?: { [key: string]: string };
  vpcConfig?: lambdafunction.LambdaFunctionVpcConfig;
  executionPolicyStatements?: iam.DataAwsIamPolicyDocumentStatement[];
  tags?: { [key: string]: string };
  logRetention?: number;
  s3Bucket: string;
  usesCodeDeploy?: boolean;
}

const DEFAULT_TIMEOUT = 5;
const DEFAULT_RETENTION = 14;

export class ApplicationVersionedLambda extends Resource {
  public readonly versionedLambda: lambdafunction.LambdaAlias;
  public lambdaExecutionRole: iam.IamRole;

  constructor(
    scope: Construct,
    name: string,
    private config: ApplicationVersionedLambdaProps
  ) {
    super(scope, name);

    this.createCodeBucket();
    this.versionedLambda = this.createLambdaFunction();
  }

  private createLambdaFunction() {
    this.lambdaExecutionRole = new iam.IamRole(this, 'execution-role', {
      name: `${this.config.name}-ExecutionRole`,
      assumeRolePolicy: this.getLambdaAssumePolicyDocument(),
    });

    const executionPolicy = new iam.IamPolicy(this, 'execution-policy', {
      name: `${this.config.name}-ExecutionRolePolicy`,
      policy: this.getLambdaExecutionPolicyDocument(),
    });

    new iam.IamRolePolicyAttachment(this, 'execution-role-policy-attachment', {
      role: this.lambdaExecutionRole.name,
      policyArn: executionPolicy.arn,
      dependsOn: [this.lambdaExecutionRole, executionPolicy],
    });

    const defaultLambda = this.getDefaultLambda();
    const lambdaConfig: lambdafunction.LambdaFunctionConfig = {
      functionName: `${this.config.name}-Function`,
      filename: defaultLambda.outputPath,
      handler: this.config.handler,
      runtime: this.config.runtime,
      timeout: this.config.timeout ?? DEFAULT_TIMEOUT,
      sourceCodeHash: defaultLambda.outputBase64Sha256,
      role: this.lambdaExecutionRole.arn,
      vpcConfig: this.config.vpcConfig,
      publish: true,
      lifecycle: {
        ignoreChanges: [
          'filename',
          'source_code_hash',
          this.shouldIgnorePublish() ? 'publish' : '',
        ].filter((v: string) => v),
      },
      tags: this.config.tags,
      environment: this.config.environment
        ? { variables: this.config.environment }
        : undefined,
    };

    const lambda = new lambdafunction.LambdaFunction(
      this,
      'lambda',
      lambdaConfig
    );

    new cloudwatch.CloudwatchLogGroup(this, 'log-group', {
      name: `/aws/lambda/${lambda.functionName}`,
      retentionInDays: this.config.logRetention ?? DEFAULT_RETENTION,
      dependsOn: [lambda],
    });

    return new lambdafunction.LambdaAlias(this, 'alias', {
      functionName: lambda.functionName,
      functionVersion: Fn.element(Fn.split(':', lambda.qualifiedArn), 7),
      name: 'DEPLOYED',
      lifecycle: {
        ignoreChanges: ['function_version'],
      },
      dependsOn: [lambda],
    });
  }

  private shouldIgnorePublish() {
    if (this.config.usesCodeDeploy !== undefined) {
      return this.config.usesCodeDeploy;
    }

    return false;
  }

  private getLambdaAssumePolicyDocument() {
    return new iam.DataAwsIamPolicyDocument(this, 'assume-policy-document', {
      version: '2012-10-17',
      statement: [
        {
          effect: 'Allow',
          actions: ['sts:AssumeRole'],
          principals: [
            {
              identifiers: ['lambda.amazonaws.com', 'edgelambda.amazonaws.com'],
              type: 'Service',
            },
          ],
        },
      ],
    }).json;
  }

  private getLambdaExecutionPolicyDocument() {
    const document = {
      version: '2012-10-17',
      statement: [
        {
          effect: 'Allow',
          actions: [
            'logs:CreateLogGroup',
            'logs:CreateLogStream',
            'logs:PutLogEvents',
            'logs:DescribeLogStreams',
          ],
          resources: ['arn:aws:logs:*:*:*'],
        },
        ...(this.config.executionPolicyStatements ?? []),
      ],
    };

    if (this.config.vpcConfig) {
      document.statement.push({
        effect: 'Allow',
        actions: [
          'ec2:DescribeNetworkInterfaces',
          'ec2:CreateNetworkInterface',
          'ec2:DeleteNetworkInterface',
          'ec2:DescribeInstances',
          'ec2:AttachNetworkInterface',
        ],
        resources: ['*'],
      });
    }

    return new iam.DataAwsIamPolicyDocument(
      this,
      'execution-policy-document',
      document
    ).json;
  }

  private getDefaultLambda() {
    const source = this.getDefaultLambdaSource();
    return new DataArchiveFile(this, 'lambda-default-file', {
      type: 'zip',
      source: [source],
      outputPath: `${source.filename}.zip`,
    });
  }

  private getDefaultLambdaSource(): DataArchiveFileSource {
    const runtime = this.config.runtime.match(/[a-z]*/)[0];
    const handler = this.config.handler.split('.');
    const functionName = handler.pop();
    const functionFilename = handler.join('.');

    let content = `exports.${functionName} = (event, context) => { console.log(event) }`;
    let filename = `${functionFilename}.js`;

    if (runtime === 'python') {
      content = `import json\ndef ${functionName}(event, context):\n\t print(event)\n\t return {'statusCode': 200, 'headers': {'dance': 'party'}, 'body': json.dumps({'electric': 'boogaloo'}), 'isBase64Encoded': False}`;
      filename = `${functionFilename}.py`;
    }

    return {
      content,
      filename,
    };
  }

  private createCodeBucket() {
    const codeBucket = new s3.S3Bucket(this, 'code-bucket', {
      bucket: this.config.s3Bucket,
      acl: 'private',
      tags: this.config.tags,
      forceDestroy: true,
    });

    new s3.S3BucketPublicAccessBlock(this, `code-bucket-public-access-block`, {
      bucket: codeBucket.id,
      blockPublicAcls: true,
      blockPublicPolicy: true,
    });
  }
}
