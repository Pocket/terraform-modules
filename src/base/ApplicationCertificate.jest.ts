import { Testing, TerraformStack } from 'cdktf';
import { ApplicationCertificate } from './ApplicationCertificate';

test('throws an error without a zone id or zone domain', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  expect(() => {
    new ApplicationCertificate(stack, 'testCert', {
      domain: 'dev.gobowling.info',
    });
  }).toThrow('You need to pass either a zone id or a zone domain');
});

test('renders a cert with a zone id', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new ApplicationCertificate(stack, 'testCert', {
    domain: 'dev.gobowling.info',
    zoneId: 'malibu',
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('renders a cert with a zone domain', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new ApplicationCertificate(stack, 'testCert', {
    domain: 'dev.gobowling.info',
    zoneId: 'gobowling.info',
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('renders a cert with tags', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new ApplicationCertificate(stack, 'testCert', {
    domain: 'dev.gobowling.info',
    zoneId: 'gobowling.info',
    tags: {
      name: 'thedude',
      hobby: 'bowling',
    },
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});
