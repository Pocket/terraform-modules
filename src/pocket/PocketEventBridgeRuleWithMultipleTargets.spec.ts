import { Testing } from 'cdktf';
import {
  PocketEventBridgeRuleWithMultipleTargets,
  PocketEventBridgeProps,
} from './PocketEventBridgeRuleWithMultipleTargets';
import {
  PocketVersionedLambda,
  PocketVersionedLambdaProps,
} from './PocketVersionedLambda';
import { LAMBDA_RUNTIMES } from '../base/ApplicationVersionedLambda';
import { SqsQueue } from '@cdktf/provider-aws/lib/sqs-queue';

describe('PocketEventBridgeRuleWithMultipleTargets', () => {
  test('renders an event bridge and multiple targets', () => {
    const synthed = Testing.synthScope((stack) => {
      const lambdaConfig: PocketVersionedLambdaProps = {
        name: 'test-lambda',
        lambda: {
          runtime: LAMBDA_RUNTIMES.PYTHON38,
          handler: 'index.handler',
        },
      };
      const testLambda = new PocketVersionedLambda(
        stack,
        'test-lambda',
        lambdaConfig
      );

      const testSqs = new SqsQueue(stack, 'test-queue', {
        name: 'Test-SQS-Queue',
      });

      const testConfig: PocketEventBridgeProps = {
        eventRule: {
          name: 'test-event-bridge-rule-multiple-targets',
          eventPattern: {
            source: ['aws.states'],
            'detail-type': ['Step Functions Execution Status Change'],
          },
        },
        targets: [
          {
            targetId: 'test-lambda-id',
            arn: 'lambda.arn',
            terraformResource: testLambda.lambda.versionedLambda,
          },
          {
            targetId: 'test-sqs-id',
            arn: testSqs.arn,
            terraformResource: testSqs,
          },
        ],
      };

      new PocketEventBridgeRuleWithMultipleTargets(
        stack,
        'test-event-bridge-for-multiple-targets-1',
        {
          ...testConfig,
          eventRule: {
            ...testConfig.eventRule,
            description: 'Test description',
          },
        }
      );
    });
    expect(synthed).toMatchSnapshot();
  });

  test('renders an event bridge and pre-existing targets', () => {
    const synthed = Testing.synthScope((stack) => {
      const testConfig: PocketEventBridgeProps = {
        eventRule: {
          name: 'test-event-bridge-rule-multiple-targets',
          eventPattern: {
            source: ['aws.states'],
            'detail-type': ['Step Functions Execution Status Change'],
          },
        },
        targets: [
          {
            targetId: 'test-lambda-id',
            arn: 'lambda.arn',
          },
          {
            targetId: 'test-sqs-id',
            arn: 'testSqs.arn',
          },
        ],
      };

      new PocketEventBridgeRuleWithMultipleTargets(
        stack,
        'test-event-bridge-for-multiple-targets-1',
        {
          ...testConfig,
          eventRule: {
            ...testConfig.eventRule,
            description: 'Test description',
          },
        }
      );
    });
    expect(synthed).toMatchSnapshot();
  });

  test('renders an event bridge rule with prevent destroy flag', () => {
    const synthed = Testing.synthScope((stack) => {
      const testConfig: PocketEventBridgeProps = {
        eventRule: {
          name: 'test-event-bridge-rule-multiple-targets',
          eventPattern: {
            source: ['aws.states'],
            'detail-type': ['Step Functions Execution Status Change'],
          },
          preventDestroy: true,
        },
        targets: [
          {
            targetId: 'test-lambda-id',
            arn: 'lambda.arn',
          },
          {
            targetId: 'test-sqs-id',
            arn: 'testSqs.arn',
          },
        ],
      };

      new PocketEventBridgeRuleWithMultipleTargets(
        stack,
        'test-event-bridge-for-multiple-targets-1',
        {
          ...testConfig,
          eventRule: {
            ...testConfig.eventRule,
            description: 'Test description',
          },
        }
      );
    });
    expect(synthed).toMatchSnapshot();
  });
});
