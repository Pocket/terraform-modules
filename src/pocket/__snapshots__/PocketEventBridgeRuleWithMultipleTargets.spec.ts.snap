// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an event bridge and multiple target 1`] = `
"{
  \\"data\\": {
    \\"archive_file\\": {
      \\"test-lambda_lambda-default-file_7A7DBB68\\": {
        \\"output_path\\": \\"index.py.zip\\",
        \\"source\\": [
          {
            \\"content\\": \\"import json\\\\ndef handler(event, context):\\\\n\\\\t print(event)\\\\n\\\\t return {'statusCode': 200, 'headers': {'dance': 'party'}, 'body': json.dumps({'electric': 'boogaloo'}), 'isBase64Encoded': False}\\",
            \\"filename\\": \\"index.py\\"
          }
        ],
        \\"type\\": \\"zip\\"
      }
    },
    \\"aws_iam_policy_document\\": {
      \\"test-lambda_assume-policy-document_D538087D\\": {
        \\"statement\\": [
          {
            \\"actions\\": [
              \\"sts:AssumeRole\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"principals\\": [
              {
                \\"identifiers\\": [
                  \\"lambda.amazonaws.com\\",
                  \\"edgelambda.amazonaws.com\\"
                ],
                \\"type\\": \\"Service\\"
              }
            ]
          }
        ],
        \\"version\\": \\"2012-10-17\\"
      },
      \\"test-lambda_execution-policy-document_D94D17B4\\": {
        \\"statement\\": [
          {
            \\"actions\\": [
              \\"logs:CreateLogGroup\\",
              \\"logs:CreateLogStream\\",
              \\"logs:PutLogEvents\\",
              \\"logs:DescribeLogStreams\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"resources\\": [
              \\"arn:aws:logs:*:*:*\\"
            ]
          }
        ],
        \\"version\\": \\"2012-10-17\\"
      }
    }
  },
  \\"resource\\": {
    \\"aws_cloudwatch_event_rule\\": {
      \\"test-event-bridge-for-multiple-targets-1_event-bridge-rule_1CCF9D7B\\": {
        \\"description\\": \\"Test description\\",
        \\"event_bus_name\\": \\"default\\",
        \\"event_pattern\\": \\"{\\\\\\"source\\\\\\":[\\\\\\"aws.states\\\\\\"],\\\\\\"detail-type\\\\\\":[\\\\\\"Step Functions Execution Status Change\\\\\\"]}\\",
        \\"name\\": \\"test-event-bridge-rule-multiple-targets-Rule\\"
      }
    },
    \\"aws_cloudwatch_event_target\\": {
      \\"test-event-bridge-for-multiple-targets-1_event-bridge-rule_event-bridge-target-test-lambda-id_DE65E71F\\": {
        \\"arn\\": \\"lambda.arn\\",
        \\"dead_letter_config\\": {
        },
        \\"depends_on\\": [
          \\"aws_lambda_alias.test-lambda_alias_CCFDFCE9\\",
          \\"aws_cloudwatch_event_rule.test-event-bridge-for-multiple-targets-1_event-bridge-rule_1CCF9D7B\\"
        ],
        \\"rule\\": \\"\${aws_cloudwatch_event_rule.test-event-bridge-for-multiple-targets-1_event-bridge-rule_1CCF9D7B.name}\\",
        \\"target_id\\": \\"test-lambda-id\\"
      },
      \\"test-event-bridge-for-multiple-targets-1_event-bridge-rule_event-bridge-target-test-sqs-id_2DE721FF\\": {
        \\"arn\\": \\"\${aws_sqs_queue.test-queue.arn}\\",
        \\"dead_letter_config\\": {
        },
        \\"depends_on\\": [
          \\"aws_sqs_queue.test-queue\\",
          \\"aws_cloudwatch_event_rule.test-event-bridge-for-multiple-targets-1_event-bridge-rule_1CCF9D7B\\"
        ],
        \\"rule\\": \\"\${aws_cloudwatch_event_rule.test-event-bridge-for-multiple-targets-1_event-bridge-rule_1CCF9D7B.name}\\",
        \\"target_id\\": \\"test-sqs-id\\"
      }
    },
    \\"aws_cloudwatch_log_group\\": {
      \\"test-lambda_log-group_ACC182B5\\": {
        \\"depends_on\\": [
          \\"aws_lambda_function.test-lambda_8915D118\\"
        ],
        \\"name\\": \\"/aws/lambda/\${aws_lambda_function.test-lambda_8915D118.function_name}\\",
        \\"retention_in_days\\": 14
      }
    },
    \\"aws_iam_policy\\": {
      \\"test-lambda_execution-policy_19C785A8\\": {
        \\"name\\": \\"test-lambda-ExecutionRolePolicy\\",
        \\"policy\\": \\"\${data.aws_iam_policy_document.test-lambda_execution-policy-document_D94D17B4.json}\\"
      }
    },
    \\"aws_iam_role\\": {
      \\"test-lambda_execution-role_9D4EE856\\": {
        \\"assume_role_policy\\": \\"\${data.aws_iam_policy_document.test-lambda_assume-policy-document_D538087D.json}\\",
        \\"name\\": \\"test-lambda-ExecutionRole\\"
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"test-lambda_execution-role-policy-attachment_8796505D\\": {
        \\"depends_on\\": [
          \\"aws_iam_role.test-lambda_execution-role_9D4EE856\\",
          \\"aws_iam_policy.test-lambda_execution-policy_19C785A8\\"
        ],
        \\"policy_arn\\": \\"\${aws_iam_policy.test-lambda_execution-policy_19C785A8.arn}\\",
        \\"role\\": \\"\${aws_iam_role.test-lambda_execution-role_9D4EE856.name}\\"
      }
    },
    \\"aws_lambda_alias\\": {
      \\"test-lambda_alias_CCFDFCE9\\": {
        \\"depends_on\\": [
          \\"aws_lambda_function.test-lambda_8915D118\\"
        ],
        \\"function_name\\": \\"\${aws_lambda_function.test-lambda_8915D118.function_name}\\",
        \\"function_version\\": \\"\${element(split(\\\\\\":\\\\\\", aws_lambda_function.test-lambda_8915D118.qualified_arn), 7)}\\",
        \\"lifecycle\\": {
          \\"ignore_changes\\": [
            \\"function_version\\"
          ]
        },
        \\"name\\": \\"DEPLOYED\\"
      }
    },
    \\"aws_lambda_function\\": {
      \\"test-lambda_8915D118\\": {
        \\"filename\\": \\"\${data.archive_file.test-lambda_lambda-default-file_7A7DBB68.output_path}\\",
        \\"function_name\\": \\"test-lambda-Function\\",
        \\"handler\\": \\"index.handler\\",
        \\"lifecycle\\": {
          \\"ignore_changes\\": [
            \\"filename\\",
            \\"source_code_hash\\"
          ]
        },
        \\"publish\\": true,
        \\"role\\": \\"\${aws_iam_role.test-lambda_execution-role_9D4EE856.arn}\\",
        \\"runtime\\": \\"python3.8\\",
        \\"source_code_hash\\": \\"\${data.archive_file.test-lambda_lambda-default-file_7A7DBB68.output_base64sha256}\\",
        \\"timeout\\": 5
      }
    },
    \\"aws_s3_bucket\\": {
      \\"test-lambda_code-bucket_177F316E\\": {
        \\"bucket\\": \\"pocket-test-lambda\\",
        \\"force_destroy\\": true
      }
    },
    \\"aws_s3_bucket_acl\\": {
      \\"test-lambda_code-bucket-acl_C73BD0DB\\": {
        \\"acl\\": \\"private\\",
        \\"bucket\\": \\"\${aws_s3_bucket.test-lambda_code-bucket_177F316E.bucket}\\"
      }
    },
    \\"aws_s3_bucket_public_access_block\\": {
      \\"test-lambda_code-bucket-public-access-block_1E7CE5AE\\": {
        \\"block_public_acls\\": true,
        \\"block_public_policy\\": true,
        \\"bucket\\": \\"\${aws_s3_bucket.test-lambda_code-bucket_177F316E.id}\\"
      }
    },
    \\"aws_sqs_queue\\": {
      \\"test-queue\\": {
        \\"name\\": \\"Test-SQS-Queue\\"
      }
    }
  }
}"
`;
